sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/CounterCell.java

```diff
@@ -0,0 +1,25 @@
+package org.apache.beam.sdk.metrics;
+
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+
+/**
+ * Tracks the current value (and delta) for a Counter metric for a specific context and bundle.
+ */
+@Experimental(Kind.METRICS)
+class CounterCell extends MetricCell<Long> {
+
+  private final AtomicLong value = new AtomicLong();
+
+  /** Increment the counter by the given amount. */
+  public void add(long n) {
+    markDirty();
```

Comment:
See latest commits -- it has been moved afterwards, and is now called "markDirtyAfterUpdate" to emphasize it should come at the end.